docs/proposals/20220209-machinepool-machines.md

```diff
@@ -0,0 +1,196 @@
+
+For example, Cluster Autoscaler needs to delete individual Machines. Without a representation in the CAPI API of individual Machines in MachinePools, projects building integrations with Cluster API would need to be specialized to delete the cloud-specific resources for each MachinePool infrastructure provider. By exposing MachinePool Machines in CAPI, Cluster Autoscaler can treat Machines built by MachineDeployments and MachinePools in a uniform fashion through CAPI APIs.
+
+As another example, currently each MachinePool infrastructure provider must implement their own node drain functionality. When MachinePool Machines become available in the CAPI API, all providers can use its existing node drain implementation.
+
+### Goals
+
+- Enable CAPI MachinePool to create Machines to represent infrastructure provider MachinePool Machines
+- Ensure that Machines in MachinePools can participate in the same behaviors as MachineDeployment Machines (node drain, machine health checks, retired node deletion)
+- Enhance `clusterctl describe` to display MachinePool Machines
+- Allow CAPI MachinePool to include deployment strategies mirroring those of MachineDeployments (future work)
+
+### Non-Goals/Future Work
+
+- Require infrastructure providers to implement MachinePool Machines. This is intended to be an "opt-in" feature, since there are existing MachinePool implementations that cannot support it. For example, `AWSManagedMachinePool` does not allow for deleting individual instances.
+- Implement Cluster Autoscaler MachinePool integration (a separate task, although we have a working POC)
+- Move the MachinePools implementation into a `exp/machinepools` package. The current code location makes its separation from `exp/addons` unclear and this would be an improvement.
+- Improve bootstrap token handling for MachinePools. It would be preferable to have a single, short-lived token for each Machine instead of the current shared, long-lived token.
+- Remove the `ProviderIDList` field from MachinePools. While the initial proposal using an `InfrastructureRefList` made that field arguably unnecessary, it is still used by cluster-autoscaler and perhaps in other integrations.
+## Proposal
+
+To create MachinePool Machines, a MachinePool in CAPI needs information about the instances or replicas associated with the provider's implementation of the MachinePool. This information is attached to the provider's MachinePool infrastructure resource as new, optional status fields `InfrastructureSelector`. This field may be populated by the infrastructure provider.
+
+```golang
+// FooMachinePoolStatus defines the observed state of FooMachinePool.
+type FooMachinePoolStatus struct {
+	// InfrastructureSelector is a label query over FooMachinePoolMachines.
+	// More info: https://kubernetes.io/docs/concepts/overview/working-with-objects/labels/#label-selectors
+	// +optional
+	InfrastructureSelector metav1.LabelSelector `json:"infrastructureSelector,omitempty"`
```

Comment:
I like this a lot better and seems less intrusive putting this info in status. Last nit: should the naming of this field be a bit more specific, for eg. `InfrastructureMachineSelector` so it's clear that's it's for MachinePool machines specifically?

Reply:
Agreed, that's a better name. 👍🏻